crates/bevy_ecs/src/system/system.rs

```diff
@@ -55,6 +48,8 @@ pub trait System: Send + Sync + 'static {
     fn apply_buffers(&mut self, world: &mut World);
     /// Initialize the system.
     fn initialize(&mut self, _world: &mut World);
+    /// Update the system's cached [`Archetypes`](crate::archetypes::Archetypes).
+    fn update_archetypes(&mut self, world: &World);
```

Comment:
I think we should rename this function to `update_archetype_component_access` (final bikeshed suggestion), since it's there to make sure the system's `Access<ArchetypeComponentId>` is up-to-date. 

We do not mutate any `Archetype` values or cache them in the system, so both this name and doc are misleading.

Also for safety and functional parity with `SystemState::get_mut`, `System::run` should do this.
```rust
fn run(&mut self, input: Self::In, world: &mut World) -> Self::Out {
    self.update_archetype_component_access(world); // verifies that `world` matches
    // SAFE: world is exclusively borrowed and the same one used to construct system
    unsafe { self.run_unsafe(input, world) }
}
```

Reply:
Yes please to both suggestions.

Reply:
SGTM. The parallel executor is currently still calling this explicitly so that it can update the metadata afterward; I'm not sure if there's a better way to do this.